rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,568 @@
+    return m.Redis.Close()
+}
+```
+
+## Custom Module configuration
+
+Custom modules can be configured using a YAML file that is loaded by the router at startup `config.yaml`. We reserve a section in the configuration file for custom modules. Each module can have its own configuration section with custom properties.
+The name of the module in the configuration file must match the name specified in the `ModuleInfo` struct.
+
+```yaml
+# config.yaml
+modules:
+  my_module:
+    value: 42
+```
+
+At module provisioning, the configuration is loaded and is marshaled into the module struct, where it can be accessed across the module lifecycle. We support all primitive types, slices, and maps in the configuration file.
+
+```go
+type MyModule struct {
+	Value uint64 `yaml:"value"`
+
+	Nested struct {
+		Property string `yaml:"property"`
+	} `yaml:"nested"`
+}
+
+func (m *MyModule) Provision(ctx *core.ModuleContext) error {
+	// Access the custom value from the configuration
+	m.Value
+}
+
+func (m *MyModule) Module() core.ModuleInfo {
+	return core.ModuleInfo{
+		Name: "myModule",
+		New:  func() core.Module { return &MyModule{} },
+	}
+}
+```
+
+## Custom Authentication and Authorization
+
+Custom modules can be used to implement custom authentication and authorization logic in the gateway. The module can intercept incoming requests and validate the user's credentials, scopes, and permissions before forwarding the request to the subgraph. The router has built-in support for JWK. The parsed token information is available in the request `req.Request.Auth` field.
+
+```go
+type MyModule struct{}
+
+var _ AuthenticationHooks = (*MyModule)(nil)
+
+func (m *MyModule) OnAuthenticate(req *core.GatewayRequest, err error) error {
+	// Authenticate the user's credentials
+	if !authenticateUser(req.Request.Auth) {
+		return core.UnauthenticatedError("Unauthenticated")
+	}
+	if !authorizeUser(req.Request.Auth) {
+		return core.UnauthenticatedError("Unauthorized")
+	}
+	return nil
+}
+```
+
+# Advanced Modules
+
+## Custom Subgraph Transport
+
+Overwrite the subgraph transport to use a custom HTTP client with retries, mTLS, timeouts, and circuit breaking.
+
+```go
+type MyModule struct {
+	Client *http.Client
+}
+
+func (m *MyModule) Provision(ctx *core.ModuleContext) error {
+	// Register your custom HTTP client as the transport for subgraph requests
+	// Can only be done in the provision method. If a second module tries to register a transport,
+	// an error is returned because the behavior is undefined.
+	
+	ctx.RegisterOriginTransport(&http.Transport{})
+	return nil
+}
+```
+
+## React to specific GraphQL Directives
+
+React to specific GraphQL directives in the subgraph schema and decide if the request should be forwarded to the subgraph. The following example checks if the `@requiresScopes` directive is present in the operation. Another example could be to check if the `@cacheControl` directive is present and set the cache control header on the response.
```

Comment:
This is very powerful.

At what specific point in the request chain will such a handler be called?
- Normalization
- Planning
- Validation
- Execution

It could make sense to allow the user to choose.

Reply:
My idea was to call them after the GraphQL operation has been fully parsed, normalized and planned. After this custom logic can be executed. The handler is also called on subgraph schema directives. The User can distinguish both cases by looking into the context e.g. when the current subgraph is set.